File: python3/construct_string_from_binary_tree.py
from .model import TreeNode

# Space : O(1)
# Time  : O(n)


class Solution:
    def tree2str(self, root: TreeNode) -> str:
        if not root:
            return ""

        l, r = "", ""
        left = self.tree2str(root.left)
        if left:
            l = "({})".format(left)

        right = self.tree2str(root.right)
        if right:
            r = "({})".format(right)

        if not l and r:
            l = "()"

        return str(root.val) + l + r



            

File: golang/construct_string_from_binary_tree.go
package main

import (
	"golang/model"
	"strconv"
)

// Space : O(1)
// Time  : O(n)

func tree2str(root *model.TreeNode) string {
    s := ""
    
    if root != nil {
        s = strconv.Itoa(root.Val)
    
        if root.Left != nil {
            s += "(" + tree2str(root.Left) + ")"
        }

        if root.Left == nil && root.Right != nil {
            s += "()"
        }
        if root.Right != nil {
            s += "(" + tree2str(root.Right) + ")"
        }
    }
    
    return s
}